hangman/lib/hangman/game.ex

```diff
@@ -0,0 +1,133 @@
+defmodule Hangman.Game do
+
+  defstruct game_state: :initializing,   # :won | :lost | :already_used | :good_guess | :bad_guess | :initializing
+            turns_left: 7,               # the number of turns left (game starts with 7)
+            letters: [],                 # a list of single character strings. If a letter in a particular
+                                         # position has been guessed, that letter will appear in `letters`. 
+                                         # Otherwise, it will be shown as an underscore
+            used: [],                    # A sorted list of the letters already guessed
+            last_guess: "",              # the last letter guessed by the player
+            word: ""                     # the secret word attempting to be guessed
+
+  def new_game() do
+    word = Dictionary.random_word
+    %Hangman.Game{
+      word: word,
+      letters: get_underscore_list(word)
+    }
+  end
+
+  def tally(game) do
+    game
+    |> Map.from_struct
+    |> Map.delete(:word)
+  end
+
+  def make_move(state = %{word: word, used: used, letters: letters}, guess) do
+    game = state 
+    |> Map.put(:last_guess, guess)
+
+    letter_in_word = word |> String.contains?(guess)
+    letter_is_used = guess in used
+    letters_unguessed = "_" in letters
+    valid_letter = String.match?(guess, ~r/^[a-z]$/)
+
+    do_guess(game, letter_in_word, letter_is_used, letters_unguessed, valid_letter)
+  end
+
+  defp get_underscore_list(word) do
+    List.duplicate("_", String.length(word))
+  end
+
+  defp get_letter_representation(letter, true) do
+    letter
+  end
+  
+  defp get_letter_representation(_, false) do
+    "_"
+  end
+
+  defp set_game_state(state, new_game_state) do
+    state
+    |> Map.put(:game_state, new_game_state)
+  end
+
+  defp add_letter_to_used(state = %{used: used, last_guess: last_guess}) do
+    state
+    |> Map.put(:used, used ++ [last_guess])
+  end
+
+  defp decrement_turns(state = %{turns_left: turns_left}) do
+    state
+    |> Map.put(:turns_left, turns_left-1)
+  end
+
+  defp do_guess(_, _, _, _, false) do
+    raise ArgumentError, message: "Invalid letter, must be a-z"
+  end
+
+  # Letter already used
+  defp do_guess(state, _, true, _, _) do
+    game = state 
+    |> set_game_state(:already_used)
+
+    { game, tally(game) }
```

Comment:
## -2 elegance

How about one more function:

~~~ elixir
defp return_with_tally(game) do
  { game, tally(game)}
end
~~~


and then use that at the end of all the other functions?